src/abstractdataframe/reshape.jl

```diff
@@ -121,169 +198,263 @@ function _stackview(df::AbstractDataFrame, measure_vars::AbstractVector{Int},
+│ 2   │ 2     │ 1     │ 1     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 2     │ 1     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 2     │ 2     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 3     │ 2     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ 3     │ 2     │ -0.172475 │ -2.81039  │
+
+julia> unstack(long, :id, :variable, :value)
+6×3 DataFrame
+│ Row │ id    │ c         │ d         │
+│     │ Int64 │ Float64?  │ Float64?  │
+├─────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ -0.172475 │ -2.81039  │
+
+julia> unstack(long, [:id, :a], :variable, :value)
+6×4 DataFrame
+│ Row │ id    │ a     │ c         │ d         │
+│     │ Int64 │ Int64 │ Float64?  │ Float64?  │
+├─────┼───────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ 1     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 2     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 2     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 3     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ 3     │ -0.172475 │ -2.81039  │
+
+julia> unstack(long, :id, :variable, :value, renamecols=x->Symbol(:_, x))
+6×3 DataFrame
+│ Row │ id    │ _c        │ _d        │
+│     │ Int64 │ Float64?  │ Float64?  │
+├─────┼───────┼───────────┼───────────┤
+│ 1   │ 1     │ 1.20649   │ -1.27628  │
+│ 2   │ 2     │ -0.917794 │ 0.940007  │
+│ 3   │ 3     │ 0.309629  │ 0.820397  │
+│ 4   │ 4     │ 1.46677   │ -1.03457  │
+│ 5   │ 5     │ 1.04339   │ -0.770464 │
+│ 6   │ 6     │ -0.172475 │ -2.81039  │
 ```
 Note that there are some differences between the widened results above.
 """
 function unstack(df::AbstractDataFrame, rowkey::ColumnIndex, colkey::ColumnIndex,
-                 value::ColumnIndex; renamecols::Function=identity)
-    refkeycol = categorical(df[!, rowkey])
-    droplevels!(refkeycol)
-    keycol = categorical(df[!, colkey])
-    droplevels!(keycol)
+                 value::ColumnIndex; renamecols::Function=identity,
+                 allowmissing::Bool=false, allowduplicates::Bool=false)
+    refkeycol = df[!, rowkey]
+    rowref, rowref_map = _unstack_preprocess_vector(refkeycol)
+    keycol = df[!, colkey]
+    colref, colref_map = _unstack_preprocess_vector(keycol)
     valuecol = df[!, value]
     return _unstack(df, index(df)[rowkey], index(df)[colkey],
-                    keycol, valuecol, refkeycol, renamecols)
+                    colref, colref_map, valuecol, rowref, rowref_map,
+                    renamecols, allowmissing, allowduplicates)
+end
+
+function _unstack_preprocess_vector(v::AbstractVector)
+    # make sure we re-allocate if v is already a PooledArray
+    vp = PooledArray{eltype(v)}(v)
+    rp = DataAPI.refpool(vp)
+    pool = similar(v, length(rp))
+    @assert firstindex(pool) == 1
+    copyto!(pool, rp)
```

Comment:
this `copyto!` is extremely slow if `v` is a `CategoricalArray` with very many levels. @nalimilan - can we do something to improve it here.

Reply:
I was thinking if we can use a generic fallback using `DataAPI.refarray`, but the problem is that we cannot, because `DataAPI.refarray` does not provide any guarantees on the values that it can hold (all standard implementations hold integers - normally starting from 1), but this is not guaranteed.

The line `copyto!(pool, rp)` is especially difficult, because in my implementation `rp` is a `Vector` of `CategoricalValue` and this is kind of hard to optimize.

@nalimilan - do you have any idea how to generically handle this? (actually in these codes it hurts us that we have `CategoricalArray` type - but the problem is that there is no generic way to unwrap values held in `CategoricalArray` unfortunately AFAICT).

Reply:
The solution I used in the grouping code is to check whether `DataAPI.refarray(vp) isa AbstractArray{Integer}` and if not bail out and just use the original array instead. You can also bail out if `firstindex(DataAPI.refpool(vp)) != 1`, but it's generally not too hard to handle this by adding `+firstindex(DataAPI.refpool(vp))-1` to `getindex` calls.

But here you call `DataAPI.refpool` and `DataAPI.refarray` on `vp`, which is always a `PooledArray`. Is that intended? In that case you can just work with the internal fields. Also why do you allocate `pool` instead of just using `rp`?

Reply:
I went to use `groupby` everywhere.